### ISSUE TITLE
Investigate Play WS Client support

### ISSUE THREAD

**Issue by gaeljw** [softwaremill/tapir#738]
Hi,

I was wondering if you are willing to accept a PR that would support Play WS as a client (in addition to sttp).

Note that:
- I don't know yet is this do-able
- I don't know yet if this makes sense. To be 100% honest I don't know sttp yet and maybe Play WS doesn't bring any value here.
- I'll be happy to contribute to this PR

My thought is that, it could be nice for Play apps to use a client which doesn't need any new additional dependencies (as Play WS is "bundled" in Play).

Any comment appreciated :)

**Comment by adamw** [comment]
Yes, this should be possible and makes sense. sttp is an API which wraps other (lower-level) http clients, so you might also have interpreters which use these http clients directly.

(sttp-client itself doesn't (yet) have a play-ws client)

Go ahead with the PR of course:)

**Comment by gaeljw** [comment]
I started to work on this but quickly came across several issues that make me think Play WS is not a good candidate for a Play Client Interpreter.

I've open a draft PR anyway for anyone that would want to have a look (#791).

What are these issues?
- PlayWS comes with two types: one for the request (`WsRequest`), one for the response (`WsResponse`), there is not a single type that defines both the request and the response of a HTTP call. This one is not a big issue, as we could imagine generating both a `I => WsRequest` function as well as a `WsResponse => DecodeResult[Either[E, O]]` (it's what I started to do...).
- Input or output in Play respectively needs a `BodyWritable`/`BodyReadable` instance, this is fine for known basic types but this create an issue when it comes to streaming body input (or output) because the type can be anything, then how to provides the `BodyWritable`/`BodyReadable` instance into Tapir?
- Multipart body seems tricky to handle with Play WS, or at least with the play-ws standalone dependency, maybe with the whole Play framework it's easier.

Considering that there is for now no demand on this feature, I won't work more on it but I'll be happy to give another try in the future if this change.

Also note that I'm far from knowing Play WS fine details, maybe someone else with more knowledge would easily solve the issues I raised.

**Comment by adamw** [comment]
Thanks a lot, even if it's only a draft! Both positive and negative investigations are useful :)

sttp is indeed at an advantage as the description of the request also handles information on how to parse the response. That's why we can translate an endpoint into a single value. Here indeed, you'd need to translate the endpoint in a tuple (as you write).

Does play-ws support any kind of streaming? If not, then you simply don't add support that capability in the interpreter and there are no problems with streams ;)

And finally - multiparts are always tricky ;)

**Comment by adamw** [comment]
if you don't specify that the interpreter supports the `Streams` capability, it won't be possible to use it with endpoints which require it. So we're safe here.

There's no capability for multipart, so that would be an ugly `IllegalArgumentException` as you write - no other choice right now :)

**Comment by gaeljw** [comment]
I made some adjustements and I think the PR is now ready for review :)

It still doesn't support Multipart, I gave it a try but it would require copying a bunch of stuff from Play framework source code that is not in Play Standalone WS lib (I don't even know if it would be okay from a license POV :shrug: ).

But it supports streaming, only with the `AkkaStreams` capability. Not sure if and how we can prevent users to use another streaming capability?

All feedbacks welcomed of course :)
